opt: GROUP BY panics on invalid ARRAY aggregations

Minimal reproduction:

`select 0, unnest(array[0]) group by 1`

panics with:

```
panic: top-level relational expression cannot have outer columns: (1) [recovered]
	panic: panic while executing 1 statements: SELECT _, unnest(ARRAY[_]) GROUP BY _; caused by top-level relational expression cannot have outer columns: (1)
```

This should gracefully fail, not crash the server.

--- a/pkg/sql/opt/optbuilder/groupby.go
+++ b/pkg/sql/opt/optbuilder/groupby.go
@@ -515,3 +515,10 @@ var aggOpLookup = map[string]opt.Operator{
 	"json_agg":   opt.JsonAggOp,
 	"jsonb_agg":  opt.JsonbAggOp,
 }
+
+func newGroupingError(name *tree.Name) error {
+	return pgerror.NewErrorf(pgerror.CodeGroupingError,
+		"column \"%s\" must appear in the GROUP BY clause or be used in an aggregate function",
+		tree.ErrString(name),
+	)
+}

--- a/pkg/sql/opt/optbuilder/scalar.go
+++ b/pkg/sql/opt/optbuilder/scalar.go
@@ -141,10 +141,7 @@ func (b *Builder) buildScalar(
 			// Non-grouping column was referenced. Note that a column that is part
 			// of a larger grouping expression would have been detected by the
 			// groupStrs checking code above.
-			panic(builderError{pgerror.NewErrorf(pgerror.CodeGroupingError,
-				"column \"%s\" must appear in the GROUP BY clause or be used in an aggregate function",
-				tree.ErrString(&t.name),
-			)})
+			panic(builderError{newGroupingError(&t.name)})
 		}
 
 		return b.finishBuildScalarRef(t, inScope, outScope, outCol, colRefs)
@@ -407,6 +404,12 @@ func (b *Builder) buildScalar(
 
 	case *srf:
 		if len(t.cols) == 1 {
+			if inGroupingContext {
+				// Non-grouping column was referenced. Note that a column that is part
+				// of a larger grouping expression would have been detected by the
+				// groupStrs checking code above.
+				panic(builderError{newGroupingError(&t.cols[0].name)})
+			}
 			return b.finishBuildScalarRef(&t.cols[0], inScope, outScope, outCol, colRefs)
 		}
 		list := make([]memo.GroupID, len(t.cols))

--- a/pkg/sql/opt/optbuilder/testdata/srfs
+++ b/pkg/sql/opt/optbuilder/testdata/srfs
@@ -908,3 +908,19 @@ project
                      │         │                   └── true [type=bool]
                      │         └── const: 1000 [type=int]
                      └── true [type=bool]
+
+# Regression test for #30412.
+build
+SELECT 0, unnest(ARRAY[0]) GROUP BY 1
+----
+error (42803): column "unnest" must appear in the GROUP BY clause or be used in an aggregate function
+
+build
+SELECT 0, unnest(ARRAY[0]) GROUP BY 1, 2
+----
+error: unnest(): generator functions are not allowed in GROUP BY
+
+build
+SELECT 0, information_schema._pg_expandarray(ARRAY[0]) GROUP BY 1
+----
+error (42803): column "x" must appear in the GROUP BY clause or be used in an aggregate function

